quint/src/runtime/impl/compilerImpl.ts

```diff
@@ -1447,6 +1487,12 @@ export class CompilerVisitor implements IRVisitor {
     // non-deterministically. Instead of modeling non-determinism,
     // we use a random number generator. This may change in the future.
     const lazyCompute = () => {
+      // on `any`, we reset the action taken as the goal is to save the last
+      // action picked in an `any` call
+      this.actionTaken = none()
+      // we also reset nondet picks since they should correspond to the action taken
+      this.nondetPicks = this.emptyNondetPicks()
```

Comment:
Not sure about this one. consider this rough example:

```
action step =
any {
   someAction,
   nondet time = oneOf(times)
   timeRelatedAction(time)
}

action timeRelatedAction(time) = 
any {
  AdvanceTime(time),
  RevertTime(time)
}
```

If I get the code correctly, it would reset the fields to None here?

Reply:
Seems I didn't understand the code correctly, because testing locally this is not true and in my example the nondet values are still stored

Reply:
I looked again and you are right, I think my comment there is misleading - I was a bit confused myself yesterday, and this is code I wrote a month ago :grimacing:. I tried removing this reset yesterday and it results in some leftovers between steps.

The thing that makes your example work well is something I missed while preparing the PR: On let-in definitions, as in
```
nondet time = oneOf(times)
timeRelatedAction(time)
```
We first evaluate the body (`timeRelatedAction(time)`) and only then add `time` to `nondetPicks`, and this is why it works. In other words, we construct the `nondetPicks` record while we are going *up* the IR tree in compilation.

This is where it happens:
https://github.com/informalsystems/quint/blob/gabriela%2Fmbt-metadata/quint/src/runtime/impl/compilerImpl.ts#L344. I'll add a comment there with this extra context. Thanks for the careful review :heart:

Reply:
Ah, nice! Thank you!